src/main/java/RecommendationsServlet.java

```diff
@@ -0,0 +1,184 @@
+    for (int i = 0; i < userHistory.size(); i++) {
+      VisualizationData currentElement = userHistory.get(i);
+      updateMap(personTypeMap, currentElement.getPersonType());
+      updateMap(actionMap, currentElement.getAction());
+      updateMap(locationMap, currentElement.getLocation());
+      updateMap(yearsMap, currentElement.getYear());
+    }
+  }
+
+  // Takes in a hashmap and returns a linkedHashMap sorted by value in descending order such that 
+  // the most frequented query types are first in the list
+  private LinkedHashMap<String, Integer> sortHashMapDescending(HashMap<String, Integer> frequencyMap) {
+    LinkedHashMap<String, Integer> reverseSortedMap = new LinkedHashMap<>();
+        frequencyMap.entrySet().stream().sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
+                .forEachOrdered(x -> reverseSortedMap.put(x.getKey(), x.getValue()));
+    return reverseSortedMap;
+  }
+
+  // Return suggested recommendations based on the history. Create hashmaps for each input for all of the queries,
+  // then create four combinations based on the most frequented fields.
+  private ArrayList<VisualizationData> getRecommendations(ArrayList<VisualizationData> userHistory, String userId) {
+    // Populate the frequency maps for each field
+    getFrequencyMaps(userHistory);
+    // Sort the frequency maps in descending order by value such that the most common searches for each field 
+    // are first eg: {worked in: 7, lived in: 10, moved to: 2} => { lived in: 10, worked in: 7, moved to: 2}
+    LinkedHashMap<String, Integer> reverseSortedPersonType = sortHashMapDescending(personTypeMap);
+    LinkedHashMap<String, Integer> reverseSortedAction = sortHashMapDescending(actionMap);
+    LinkedHashMap<String, Integer> reverseSortedLocation = sortHashMapDescending(locationMap);
+    LinkedHashMap<String, Integer> reverseSortedYear = sortHashMapDescending(yearsMap);
+    // Get the keyset from each frequency map with the guaranteed order from sorting
+    Set<String> personTypeKeys = reverseSortedPersonType.keySet();
+    Set<String> actionKeys = reverseSortedAction.keySet();
+    Set<String> locationKeys = reverseSortedLocation.keySet();
+    Set<String> yearKeys = reverseSortedYear.keySet();
+    ArrayList<VisualizationData> recommendationList = new ArrayList<VisualizationData>();
+    // Iterate through all fields and find combinations that the user hasn't made yet
+    // and return the top 3.
+    for (String personType : personTypeKeys) {
+      for(String action : actionKeys) {
+        for (String location : locationKeys) {
+          for (String year : yearKeys) {
+            VisualizationData recommendation = new VisualizationData(userId, personType, action, location, year);
+             if (isRecommendationValid(recommendation, userHistory)) {
+               recommendationList.add(recommendation);
+               if (recommendationList.size() > 3) {
```

Comment:
you can make the number of recommendations be a parameter and you can return the requested number of recommendations and padding with random recommendations if necessary.

Reply:
The three of us decided that we should only show five at a time, so I've updated getRecommendations to only return five.